site/docs/plugins/conversations.md

```diff
@@ -7,1190 +7,1507 @@ next: false
 ```
 
-As soon as the user sends `/start` to the bot, the conversation will be entered.
-The current context object is passed as the second argument to the conversation builder function.
-For example, if you start your conversation with `await ctx.reply(ctx.message.text)`, it will contain the update that contains `/start`.
+Escaping a part of your code via `conversation.external` signals to the plugin that this part of the code should be skipped during replays.
+The return value of the wrapped code is stored by the plugin and reused during subsequent replays.
+In the above example, this prevents repeated database access.
 
-::: tip Change the Conversation Identifier
-By default, you have to pass the name of the function to `ctx.conversation.enter()`.
-However, if you prefer to use a different identifier, you can specify it like so:
+USE `conversation.external` when you ...
 
-```ts
-bot.use(createConversation(greeting, "new-name"));
-```
+- read or write to files, databases/sessions, the network, or global state,
+- call `Math.random()` or `Date.now()`,
+- perform API calls on `bot.api` or other independent instances of `Api`.
+
+DO NOT USE `conversation.external` when you ...
+
+- call `ctx.reply` or other [context actions](../guide/context#available-actions),
+- call `ctx.api.sendMessage` or other methods of the Bot API via `ctx.api`.
 
-In turn, you can enter the conversation with it:
+The conversations plugin provides a few convenience methods around `conversation.external`.
+This not only simplifies using `Math.random()` and `Date.now()`, but it also simplifies debugging by providing a way to suppress logs during a replay.
 
 ```ts
-bot.command("start", (ctx) => ctx.conversation.enter("new-name"));
+// await conversation.external(() => Math.random());
+const rnd = await conversation.random();
+// await conversation.external(() => Date.now());
+const now = await conversation.now();
+// await conversation.external(() => console.log("abc"));
+await conversation.log("abc");
 ```
 
-:::
+How can `conversation.wait` and `conversation.external` recover the original values when a replay happens?
+The plugin has to somehow remember this data, right?
+
+Yes.
+
+### Conversations Store State
+
+Two types of data are being stored in a database.
+By default, it uses a lightweight in-memory database that is based on a `Map`, but you can [use a persistent database](#persisting-conversations) easily.
+
+1. The conversations plugin stores all updates.
+2. The conversations plugin stores all return values of `conversation.external` and the results of all API calls.
+
+This is not an issue if you only have a few dozen updates in a conversation.
+(Remember that during long polling, every call to `getUpdates` retrieves up to 100 updates, too.)
 
-In total, your code should now roughly look like this:
+However, if your conversation never exits, this data will accumulate and slow down your bot.
+**Avoid infinite loops.**
```

Comment:
Which part needs an example? Is it unclear what an infinite loop is?

Reply:
An example of a conversation that never exits. 

Maybe it's just my lack of understanding. I should reread the document from the start to grasp it better since I continued from where I left off and forgot all the details. Lol.

Reply:
The example would then be
```ts
while (true)
  await conversation.wait()
```